debezium-server-kafka/src/main/java/io/debezium/server/kafka/KafkaChangeConsumer.java

```diff
@@ -81,31 +91,44 @@ void stop() {
     public void handleBatch(final List<ChangeEvent<Object, Object>> records,
                             final RecordCommitter<ChangeEvent<Object, Object>> committer)
             throws InterruptedException {
-        final CountDownLatch latch = new CountDownLatch(records.size());
+        List<Future<RecordMetadata>> futures = new ArrayList<>();
         for (ChangeEvent<Object, Object> record : records) {
             try {
                 LOGGER.trace("Received event '{}'", record);
                 Headers headers = convertKafkaHeaders(record);
 
                 String topicName = streamNameMapper.map(record.destination());
-                producer.send(new ProducerRecord<>(topicName, null, null, record.key(), record.value(), headers), (metadata, exception) -> {
-                    if (exception != null) {
-                        LOGGER.error("Failed to send record to {}:", topicName, exception);
-                        throw new DebeziumException(exception);
-                    }
-                    else {
-                        LOGGER.trace("Sent message with offset: {}", metadata.offset());
-                        latch.countDown();
-                    }
-                });
+                Future<RecordMetadata> recordMetadataFuture = producer.send(new ProducerRecord<>(topicName, null, null, record.key(), record.value(), headers),
+                        (metadata, exception) -> {
+                            if (exception != null) {
+                                LOGGER.error("Failed to send record to {}:", topicName, exception);
+                                throw new DebeziumException(exception);
+                            }
+                            else {
+                                LOGGER.trace("Sent message with offset: {}", metadata.offset());
+                            }
+                        });
+                futures.add(recordMetadataFuture);
                 committer.markProcessed(record);
```

Comment:
@PlugaruT I am sorry I completely forgot to point you to https://issues.redhat.com/browse/DBZ-7575?focusedId=25342060&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25342060

Generally this markProcessed needs to be removed here and moved to the following loop that processes the futures.
And this brings one more issue - when futures are used, are the records delivered in sequence? If not I see two problems
1) Two records for the same PK can be delivered in an incorrect order
2) Theprocessing mark is not done in correct order which I belive can cause issue with offsets

Reply:
I actually have been thinking about this, but then got confused as you said one thing, and Vojtech another 😅 No probs with that. 

Reading the `producer.send` [docs](https://kafka.apache.org/23/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send-org.apache.kafka.clients.producer.ProducerRecord-), the way we use it here, with a callback, seems that it's a non-blocking call and from my understanding, the batching and ordering is done by the producer itself since it has some buffering logic. 

Now that I think about it, maybe it's worth actually reverting this change and going back to just add `latch.countDown();` in the callback itself and that would be it. 
Let me know what you think.

Reply:
In fact, is not the real issue that the code should be like
```
                producer.send(new ProducerRecord<>(topicName, null, null, record.key(), record.value(), headers), (metadata, exception) -> {
                    if (exception != null) {
                        LOGGER.error("Failed to send record to {}:", topicName, exception);
                        throw new DebeziumException(exception);
                    }
                    else {
                        LOGGER.trace("Sent message with offset: {}", metadata.offset());
                        latch.countDown();
                    }
                }).get();
```
Note the `get()` at the end. In this case if the exception is thrown it is passed up to the call and the Server will terminate. In this case it still makes sense so keep the new option, just either call get, or get with timeout. WDYT?

Reply:
Yes, makes sense to me. 
I applied the changes. Let me know if it works. Thanks